### PR TITLE
fix(pubsub): wait for successful receipt modack when exactly once delivery enabled

### PR DESCRIPTION
**Pull request by NivedhaSenthil**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [ ] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [ ] Update code documentation if necessary.

closes: #20899

### CODE REVIEW

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb [hunk 1]
```diff
@@ -295,6 +300,19 @@ def background_run
...
+          def create_receipt_modack_for_eos received_messages
+            received_messages.each do |rec_msg_grpc|
+              callback = proc do |result|
+                if result.succeeded?
+                  synchronize { @inventory.add rec_msg_grpc }
+                  rec_msg = ReceivedMessage.from_grpc rec_msg_grpc, self
+                  register_callback rec_msg
+                end
+              end
+              @subscriber.buffer.modify_ack_deadline @subscriber.deadline, [rec_msg_grpc.ack_id], callback
```
Comment (maheshgattani) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb:312]: I can't tell where the temp failures are retried. Can you help me follow, maybe a link to the code that will eventually trigger a retry for this ack id?

Comment (NivedhaSenthil) [reply google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb:312]: handling of errors happen [here](https://github.com/googleapis/google-cloud-ruby/blob/9410d03deeaff08b605302b8937248a2b82d7995/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb#L149)

**Review by maheshgattani — APPROVED** [review]

File: google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb [hunk 2]
```diff
@@ -78,6 +78,9 @@ def stub.acknowledge subscription:, ack_ids:
     stub = StreamingPullStub.new response_groups
     called = false  
     def stub.modify_ack_deadline subscription:, ack_ids:, ack_deadline_seconds:
+      if @modify_ack_deadline_requests.count == 0
+        return  @modify_ack_deadline_requests << [subscription, ack_ids.sort, ack_deadline_seconds]
```
Comment (dazuma) [inline google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb:82]: nit: extra space. Same in 3 other cases below.

**Review by dazuma — APPROVED** [review]

File: google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb (hunk 2 above)
Comment (NivedhaSenthil) [reply google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb]: Fixed 👍

**Pull request merged**